File: example/ios/WireframeGoldenTests/WireframeGoldenFormat.swift
import Foundation
import MixpanelSessionReplay
import XCTest

/// Golden ("snapshot") assertion for the `mp_wireframe` element list.
///
/// The serialized **format** is byte-for-byte identical to the Android, Flutter and iOS SDK
/// suites, so a reviewer can eyeball cross-platform parity:
///
/// - 2-space indent, arrays expanded one value per line
/// - key order: `role`, `text`, `bounds`, `maskDecision`
/// - `text` is the JSON literal `null` when the element carries no text
/// - no trailing newline, byte-exact string comparison
///
/// A missing golden is auto-created (written, announced, passes), so authoring a case is a
/// single run. Goldens are resolved through `#filePath` rather than the bundle, because they
/// belong to the source tree, not to the app that happens to host the test.
enum WireframeGoldenFormat {

  private static var goldenDir: URL {
    URL(fileURLWithPath: #filePath)
      .deletingLastPathComponent()
      .appendingPathComponent("Golden")
  }

  static func assertGolden(
    _ capture: WireframeGoldenHarness.Capture,
    _ golden: String,
    file: StaticString = #filePath,
    line: UInt = #line
  ) {
    assertText(
      maskJson(capture.maskFrames),
      golden.replacingOccurrences(of: ".json", with: ".masks.json"),
      file: file, line: line)
    assertText(
      json(viewport: capture.viewport, elements: capture.elements), golden,
      file: file, line: line)
  }

  /// Writes or compares one golden file.
  private static func assertText(
    _ actual: String, _ golden: String, file: StaticString, line: UInt
  ) {
    let url = goldenDir.appendingPathComponent(golden)

    guard let existing = try? String(contentsOf: url, encoding: .utf8) else {
      try? FileManager.default.createDirectory(
        at: goldenDir, withIntermediateDirectories: true)
      do {
        try actual.write(to: url, atomically: true, encoding: .utf8)
        print("📸 Created wireframe golden: \(golden) -> \(url.path)")
      } catch {
        XCTFail("could not write golden \(golden): \(error)", file: file, line: line)
      }
      return
    }

    XCTAssertEqual(
      existing, actual,
      """
      Wireframe golden mismatch for \(golden).
      If this change is intended, delete \(url.path) and re-run to regenerate.
      """,
      file: file, line: line)
  }

  /// Mask rectangles, in the same shape `android-goldens/` writes.
  private static func maskJson(_ frames: [[Int]]) -> String {
    if frames.isEmpty { return "{\n  \"maskFrames\": []\n}" }
    var out = "{\n  \"maskFrames\": [\n"
    for (i, f) in frames.enumerated() {
      out += "    { \"bounds\": [\(f[0]), \(f[1]), \(f[2]), \(f[3])] }"
      out += (i == frames.count - 1 ? "\n" : ",\n")
    }
    return out + "  ]\n}"
  }

  // MARK: - Serialization

  private static func json(
    viewport: [Int], elements: [WireframeGoldenHarness.Element]
  ) -> String {
    var out = "{\n"
    out += "  \"viewport\": [\n"
    for (i, v) in viewport.enumerated() {
      out += "    \(v)" + (i == viewport.count - 1 ? "\n" : ",\n")
    }
    out += "  ],\n"
    if elements.isEmpty {
      out += "  \"elements\": []\n"
    } else {
      out += "  \"elements\": [\n"
      for (i, e) in elements.enumerated() {
        out += "    {\n"
        out += "      \"role\": \(quote(e.role)),\n"
        out += "      \"text\": \(e.text.map(quote) ?? "null"),\n"
        out += "      \"bounds\": [\n"
        out += "        \(e.bounds[0]),\n"
        out += "        \(e.bounds[1]),\n"
        out += "        \(e.bounds[2]),\n"
        out += "        \(e.bounds[3])\n"
        out += "      ],\n"
        out += "      \"maskDecision\": \(quote(e.maskDecision))\n"
        out += (i == elements.count - 1 ? "    }\n" : "    },\n")
      }
      out += "  ]\n"
    }
    out += "}"
    return out
  }

  private static func quote(_ s: String) -> String {
    var out = "\""
    for ch in s.unicodeScalars {
      switch ch {
        case "\"": out += "\\\""
        case "\\": out += "\\\\"
        case "\n": out += "\\n"
        case "\r": out += "\\r"
        case "\t": out += "\\t"
        default:
          if ch.value < 0x20 {
            out += String(format: "\\u%04x", ch.value)
          } else {
            out.unicodeScalars.append(ch)
          }
      }
    }
    return out + "\""
  }
}

File: android-goldens/src/test/java/com/mixpanelreactnativesessionreplay/goldens/WireframeGoldenFormat.kt
package com.mixpanelreactnativesessionreplay.goldens

import com.mixpanel.android.sessionreplay.wireframe.WireframeElement
import org.junit.Assert.assertEquals
import java.io.File

/**
 * Golden ("snapshot") assertion for the `mp_wireframe` element list.
 *
 * The serialized **format** is byte-for-byte identical to the Android SDK's own golden harness
 * and to the Flutter and iOS suites, so a reviewer can eyeball cross-platform parity. Only the
 * coordinates differ per platform, which is why the files are not shared:
 *  - 2-space indent, arrays expanded one value per line
 *  - key order: `role`, `text`, `bounds`, `maskDecision`
 *  - `text` is the JSON literal `null` when the element carries no text
 *  - no trailing newline
 *  - byte-exact string comparison
 *
 * Unlike the instrumented harness this runs on the JVM with the host source tree writable, so a
 * missing golden is **auto-created** (written, announced, passes) exactly as the Robolectric
 * harness does — authoring a new case is a single run.
 *
 * Coordinates here come from layoutlib rather than a device, so they are pinned by the
 * `DeviceConfig` each test declares, not by whatever emulator happened to be attached.
 */
internal object WireframeGoldenFormat {

    private const val GOLDEN_RELATIVE_PATH = "src/test/golden/wireframe"

    private val goldenDir: File
        get() = File(File(System.getProperty("user.dir") ?: "."), GOLDEN_RELATIVE_PATH)

    /**
     * Asserts the element list, and the mask rectangles alongside it in `<name>.masks.json`.
     *
     * The companion file exists because masking is otherwise invisible in cases where there was
     * no text to remove — a masked, unlabelled image produces the same element as an unmasked
     * one. Same shape as the iOS SDK's `.masks.json` files.
     */
    fun assertGolden(capture: ReactNativePaparazziHarness.Capture, golden: String) {
        assertText(maskGoldenJson(capture.maskFrames), golden.removeSuffix(".json") + ".masks.json")
        val actual = wireframeGoldenJson(capture.viewport, capture.elements)
        assertText(actual, golden)
    }

    private fun assertText(actual: String, golden: String) {
        val file = File(goldenDir, golden)
        if (!file.exists()) {
            file.parentFile?.mkdirs()
            file.writeText(actual)
            println("📸 Created wireframe golden: $golden -> ${file.absolutePath}")
            return
        }
        assertEquals(
            "Wireframe golden mismatch for $golden.\n" +
                "If this change is intended, delete ${file.path} and re-run to regenerate.",
            file.readText(),
            actual
        )
    }

    private fun maskGoldenJson(frames: List<android.graphics.Rect>): String {
        if (frames.isEmpty()) return "{\n  \"maskFrames\": []\n}"
        val sb = StringBuilder("{\n  \"maskFrames\": [\n")
        frames.forEachIndexed { i, rect ->
            sb.append("    { \"bounds\": [")
                .append(rect.left).append(", ").append(rect.top).append(", ")
                .append(rect.width()).append(", ").append(rect.height())
                .append("] }")
                .append(if (i == frames.lastIndex) "\n" else ",\n")
        }
        sb.append("  ]\n}")
        return sb.toString()
    }

    // ---- Serialization (byte-compatible with the other Android harnesses + Flutter/iOS) --------

    private fun wireframeGoldenJson(viewport: List<Int>, elements: List<WireframeElement>): String {
        val sb = StringBuilder()
        sb.append("{\n")
        sb.append("  \"viewport\": [\n")
        viewport.forEachIndexed { i, v ->
            sb.append("    ").append(v).append(if (i == viewport.lastIndex) "\n" else ",\n")
        }
        sb.append("  ],\n")
        if (elements.isEmpty()) {
            sb.append("  \"elements\": []\n")
        } else {
            sb.append("  \"elements\": [\n")
            elements.forEachIndexed { index, e ->
                appendElement(sb, e)
                sb.append(if (index == elements.lastIndex) "    }\n" else "    },\n")
            }
            sb.append("  ]\n")
        }
        sb.append("}")
        return sb.toString()
    }

    private fun appendElement(sb: StringBuilder, e: WireframeElement) {
        sb.append("    {\n")
        sb.append("      \"role\": ").append(jsonString(e.type.wireName())).append(",\n")
        sb.append("      \"text\": ").append(e.text?.let(::jsonString) ?: "null").append(",\n")
        sb.append("      \"bounds\": [\n")
        sb.append("        ").append(e.x).append(",\n")
        sb.append("        ").append(e.y).append(",\n")
        sb.append("        ").append(e.w).append(",\n")
        sb.append("        ").append(e.h).append("\n")
        sb.append("      ],\n")
        sb.append("      \"maskDecision\": ").append(jsonString(e.maskDecision.name)).append("\n")
    }

    private fun jsonString(s: String): String {
        val sb = StringBuilder(s.length + 2)
        sb.append('"')
        for (ch in s) {
            when (ch) {
                '"' -> sb.append("\\\"")
                '\\' -> sb.append("\\\\")
                '\n' -> sb.append("\\n")
                '\r' -> sb.append("\\r")
                '\t' -> sb.append("\\t")
                else -> if (ch.code < 0x20) sb.append("\\u%04x".format(ch.code)) else sb.append(ch)
            }
        }
        sb.append('"')
        return sb.toString()
    }
}
